File: quests/chapter11/java/Shortener.java
// Chapter 11: Tower of Constructs - URL Shortener
// Your task: Implement Base62 encoding and decoding for URL shortening

public class Shortener {
    // Base62 character set: 0-9, A-Z, a-z
    static final String BASE62_CHARS = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";

    /**
     * Encode a numeric ID to a Base62 string
     *
     * Base62 encoding uses 62 characters (0-9, A-Z, a-z) to represent numbers.
     * This is commonly used in URL shorteners because:
     * - It's URL-safe (no special characters)
     * - It's compact (62^6 = 56+ billion unique codes with just 6 characters)
     * - It's human-readable
     *
     * TODO: Implement Base62 encoding
     * - Handle the special case when id is 0
     * - Use modulo 62 to get each digit
     * - Use integer division to move to the next digit
     * - Build the result string from least significant to most significant digit
     *
     * Example:
     *   0 -> "0"
     *   1 -> "1"
     *   62 -> "10"
     *   3844 -> "100"
     *
     * @param id The numeric ID to encode (non-negative)
     * @return The Base62 encoded string
     */
    public static String base62Encode(long id) {
        // TODO: Implement Base62 encoding
        // Hint: Similar to converting decimal to any other base
        // Use StringBuilder to build the result efficiently
        throw new UnsupportedOperationException("Not implemented yet");
    }

    /**
     * Decode a Base62 string back to a numeric ID
     *
     * This reverses the encoding process to retrieve the original number.
     *
     * TODO: Implement Base62 decoding
     * - Iterate through each character in the string
     * - Find the index of each character in BASE62_CHARS
     * - Multiply accumulated result by 62 and add the character value
     *
     * Example:
     *   "0" -> 0
     *   "1" -> 1
     *   "10" -> 62
     *   "100" -> 3844
     *
     * @param shortCode The Base62 encoded string
     * @return The decoded numeric ID
     * @throws IllegalArgumentException if the string contains invalid characters
     */
    public static long base62Decode(String shortCode) {
        // TODO: Implement Base62 decoding
        // Hint: Process each character from left to right
        // result = result * 62 + charValue
        throw new UnsupportedOperationException("Not implemented yet");
    }
}

File: quests/chapter11/go/shortener.go
// Chapter 11: Tower of Constructs - URL Shortener
// Design a scalable URL shortening system with base62 encoding

package main

// Base62Chars contains the character set for base62 encoding
// 0-9 (10) + A-Z (26) + a-z (26) = 62 characters
const Base62Chars = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"

// Base62Encode converts a number to a base62 string
// Used for generating short codes from database IDs
// Example: 0 -> "0", 62 -> "10", 3844 -> "100"
func Base62Encode(id int64) string {
	// TODO: Convert number to base62 string
	// - Handle edge case where id is 0 (return "0")
	// - Build result by repeatedly dividing by 62
	// - Use Base62Chars to map remainders to characters
	// - Remember to reverse or prepend characters
	return ""
}

// Base62Decode converts a base62 string back to a number
// Used for looking up the original ID from a short code
// Example: "0" -> 0, "10" -> 62, "100" -> 3844
func Base62Decode(shortCode string) int64 {
	// TODO: Convert base62 string to number
	// - Process each character from left to right
	// - Find the index of each character in Base62Chars
	// - Multiply running total by 62 and add the index
	return 0
}
